test/test_rake_file_utils.rb

```diff
@@ -287,6 +287,110 @@ def test_sh_show_command
+    command = build_command string, output, error_message
+    verbose true
+    actual = capture_subprocess_io do
+      sh(command){|ok, status|}
+    end
+    assert_output_equal actual, command, output, error_message
+  end
+
+  def assert_output_equal(actual, command, output, error_message)
+    # Unix standard-output, standard-error:
+    expected = [output, "#{command}\n#{error_message}"]
+    exp = standard_error_sort expected
+    act = standard_error_sort actual
+    assert_equal exp, act
+  end
+  private :assert_output_equal
+
+  def assert_sh_echoes_it_fully(string, output, error_message)
+    command = build_command string, output, error_message
+    # For testing purposes, we needn't match other programs' various
+    # error exit status code numbers, just Ruby's:
+    expected_error_message = "Command failed with status (#{error_exit_status}): [#{command}]"
+    ex = nil # Predefine to survive the block.
+    capture_subprocess_io do
+      ex = assert_raises RuntimeError do
+        sh command
+      end
+    end
+    assert_equal expected_error_message, ex.message
+  end
+  private :assert_sh_echoes_it_fully
+
+  def assert_sh_prints_its_full_output(string, output, error_message)
+    command = build_command string, output, error_message
+    actual = capture_subprocess_io do
+      assert_raises RuntimeError do
+        sh command
+      end
+    end
+    assert_output_equal actual, command, output, error_message
+  end
+  private :assert_sh_prints_its_full_output
+
+  def build_command(string, output, error_message)
+    clause_string = string.empty? ? '' : "v='#{string}';"
+    clause_stdout = output.empty? ? '' : "puts '#{output}';"
+    clause_stderr = error_message.empty? ? '' : "STDERR.puts '#{error_message}';"
+    clause_exit = "exit false"
+    "ruby -e\"#{clause_string}#{clause_stdout}#{clause_stderr}#{clause_exit}\""
+  end
+  private :build_command
+
+  def error_exit_status
+    @error_exit_status ||= begin
+      # TODO: over all Rubies and operating systems, determine whether
+      # Ruby's (Kernel.exit false) produces the same Unix exit-status
+      # value (perhaps 1)?
+      command = "ruby -e'exit false'"
+      sh command do |ok, status|
+        flunk if ok
+        @error_exit_status = status.exitstatus
+      end     
+      @error_exit_status
+    end     
+  end
+  private :error_exit_status
+
+  def standard_error_sort(both)
+    flunk unless 2 == both.length
+    stdout = both.first.strip
+    # Currently, the actual order varies among operating systems (e.g.,
+    # between Windows and Linux), so tests will fail unless we sort
+    # them:
+    stderr = both.last.split("\n").map(&:strip).reject(&:empty?).sort
+    [stdout, stderr]
+  end
+  private :standard_error_sort
+
+  def string_and_output_and_error_long() [('1234567890' * 10), 'Some output', 'Some error'] end
+  private :string_and_output_and_error_long
+
+  def string_and_output_and_error_short() ['', '', ''] end
```

Comment:
I'd rather these be duplicated inline in the setup for each test [per this comment](https://github.com/ruby/rake/pull/147/files#r71949569).  Down here in separate methods it's hard to figure out where they get used.